File: 2024/03/go/main.go
package main

import (
	"fmt"
	"io"
	"os"
	"regexp"
	"strconv"
	"strings"
)

type memory string

func (m memory) solve() (do int, dont int) {
	reg := regexp.MustCompile(`(mul\(\d+,\d+\))|(do\(\))|(don't\(\))`)

	matches := reg.FindAll([]byte(m), -1)
	isDo := true
	for _, match := range matches {
		matchStr := string(match)

		if matchStr == "do()" {
			isDo = true
		} else if matchStr == "don't()" {
			isDo = false
		} else {
			matchStr = matchStr[4:len(matchStr)-1]
			numsStr := strings.Split(matchStr, ",")

			num1, _ := strconv.Atoi(numsStr[0])
			num2, _ := strconv.Atoi(numsStr[1])
			if isDo {
				do += num1 * num2
			} else {
				dont += num1 * num2
			}
		}
	}

	return
}

var data memory

func main() {
	buf, _ := io.ReadAll(os.Stdin)
	data = memory(buf)

	do, dont := data.solve()
	P1 := do + dont
	P2 := do

	fmt.Printf("P1: %d\n", P1)
	fmt.Printf("P2: %d\n", P2)
}

File: 2024/03/typescript/main.ts
import { readFileSync } from "fs";

class Mul {
    data: string

    constructor(m: string) {
        this.data = m
    }

    multiply(): number {
        const [sn1, sn2] = this.data.slice(4, -1).split(",")
        const [n1, n2] = [parseInt(sn1), parseInt(sn2)]
        return n1 * n2
    }
}


class Memory {
    data: string
    re: RegExp

    constructor(d: string) {
        this.data = d
        this.re = /(mul\(\d+\,\d+\))|(do\(\))|(don't\(\))/g

        console.log(this.re.source)
    }

    solve(): Array<[boolean, number]> {
        const matches = this.data.match(this.re)
        let count: boolean = true

        const res: Array<[boolean, number]> = []
        for (const m of matches as RegExpMatchArray) {
            if (m === "do()") {
                count = true
            } else if (m === "don't()") {
                count = false
            } else {
                const val = new Mul(m).multiply()
                res.push([count, val])
            }
        }

        return res
    }
}

const data = new Memory(readFileSync(0, { encoding: "utf-8" }))
const res = data.solve()

let P1: number = 0
for (const [count, val] of res) {
    P1 += val
}

console.log(`P1: ${P1}`)

let P2: number = 0
for (const [count, val] of res) {
    if (count) {
        P2 += val
    }
}

console.log(`P2: ${P2}`)
